### DISCUSSION TITLE
Kernel handling in mulit chat/service solution

### DISCUSSION THREAD

**Discussion by akordowski** [microsoft/semantic-kernel#8104] [Q&A] [answered]
I am currently working on a multi chat/service solution with the Semantic Kernel and I wonder what is the best practice to handle the Kernel in such a scenario. As each chat could have its own unique configuation should the Kernel be used as singleton or a new instance for each chat?

Here an (very simplified) example:

```csharp
var kernel = new KernelBuilder().Build();

var chat1 = kernel.GetRequiredService<IChatCompletionService>();
kernel.ImportPluginFromObject("abc");
var chat1Response = chat1.GetChatMessageContentsAsync(..., kernel);

var chat2 = kernel.GetRequiredService<IChatCompletionService>();
kernel.ImportPluginFromObject("def");
var chat2Response = chat2.GetChatMessageContentsAsync(..., kernel);
```

In this example `chat1` should have access only to the `abc` plugin and the `chat2` to the `def` plugin. But I guess both chats would have access to both plugins as the `Kernel` is an singelton instance. 

I would guess that in a mutli chat scenario I should create a new `Kernel` instance for each chat and configure it respectively. Am I right?

**Comment by dmytrostruk** [comment, marked as answer]
@akordowski I think you can create a new `Kernel` instance for each chat and configure it respectively. We have `AddKernel` extension method for `IServiceCollection`, where we register it as Transient for exactly the same purpose, so in different places in your application you can configure your Kernel in unique way by adding different set of plugins, filters, data etc:
https://github.com/microsoft/semantic-kernel/blob/f9247dab0c588dca5fb6422fe6c1b4754e752136/dotnet/src/SemanticKernel.Abstractions/Services/KernelServiceCollectionExtensions.cs#L22-L29
